core/src/main/java/com/linecorp/armeria/client/backoff/Backoff.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.backoff;
+
+import static com.linecorp.armeria.client.backoff.FixedBackoff.NO_DELAY;
+
+/**
+ * Control back off between attemtps in a single retry operation.
+ */
+@FunctionalInterface
+public interface Backoff {
+    /**
+     * Returns a {@link Backoff} that provides zero interval.
+     */
+    static Backoff withoutDelay() {
+        return NO_DELAY;
+    }
+
+    /**
+     * Returns a {@link Backoff} that provides a fixed interval between two attempts.
+     */
+    static Backoff fixed(long intervalMillis) {
+        return new FixedBackoff(intervalMillis);
+    }
+
+    /**
+     * Returns a {@link Backoff} that provides an interval (that increases exponentially) between two attempts.
+     */
+    static Backoff exponential(long minIntervalMillis, long maxIntervalMillis) {
+        return exponential(minIntervalMillis, maxIntervalMillis, 2.0f);
+    }
+
+    /**
+     * Returns a {@link Backoff} that provides an interval (that increases exponentially) between two attempts.
+     */
+    static Backoff exponential(long minIntervalMillis, long maxIntervalMillis, float multiplier) {
+        return new ExponentialBackoff(minIntervalMillis, maxIntervalMillis, multiplier);
+    }
+
+    /**
+     * Returns a {@link Backoff} that provides a random interval between two attempts.
+     */
+    static Backoff random(long minJitterMillis, long maxJitterMillis) {
+        return new RandomBackoff(minJitterMillis, maxJitterMillis);
+    }
+
+    /**
+     * Returns a new back off value in milliseconds.
+     */
+    long nextIntervalMillis();
```

Comment:
How about:

- removing `maxRetries()`,
- adding `int numAttemptsSoFar` to `nextIntervalMillis()` and
- letting it return a negative value to stop retrying?

e.g.

```java
(numAttemptsSoFar) -> { // numAttemptsSoFar starts from 1
    if (numAttemptsSoFar == 3) {
        return -1; // up to three calls
    }
    ...
}
```